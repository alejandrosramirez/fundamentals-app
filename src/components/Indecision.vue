<template>
	<img v-if="image" :src="image" alt="image" />

	<div class="bg-dark"></div>

	<div class="indecision-container">
		<input type="text" placeholder="Haz una pregunta..." v-model="question" />
		<p>Recuerda terminar con un '?'</p>

		<div v-if="isValidQuestion">
			<h2>{{ question }}</h2>
			<h1 v-if="answer">{{ answer }}</h1>
		</div>
	</div>
</template>

<script>
// https://yesno.wtf/api
export default {
	name: "Indecision",
	data() {
		return {
			question: "",
			answer: null,
			image: null,
			isValidQuestion: false,
		};
	},
	methods: {
		async getAnswer() {
			try {
				this.answer = "Pensando...";

				const response = await fetch("https://yesno.wtf/api");

				const { answer, image } = await response.json();

				this.answer = answer === "yes" ? "Si!" : "No!";
				this.image = image;
			} catch (error) {
				this.answer = "Error";
				this.image = null;
			}
		}
	},
	watch: {
		question(value, oldValue) {
			this.isValidQuestion = false;

			console.log({ value });

			if (!value.includes("?")) {
				return;
			}

			this.isValidQuestion =  true;

			this.getAnswer();
		}
	},
};
</script>

<style scoped>
img,
.bg-dark {
	height: 100vh;
	left: 0px;
	max-height: 100%;
	max-width: 100%;
	position: fixed;
	top: 0px;
	width: 100vw;
}

.bg-dark {
	background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
	position: relative;
	z-index: 99;
}

input {
	width: 250px;
	padding: 10px 15px;
	border-radius: 5px;
	border: none;
}
input:focus {
	outline: none;
}

p {
	color: white;
	font-size: 20px;
	margin-top: 0px;
}

h1,
h2 {
	color: white;
}

h2 {
	margin-top: 150px;
}
</style>
