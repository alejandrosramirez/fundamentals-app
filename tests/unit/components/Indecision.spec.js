import { shallowMount } from "@vue/test-utils";

import Indecision from "@/components/Indecision.vue";

describe("Pruebas en <Indecision />", () => {
	let wrapper;
	let clgSpy;
	global.fetch = jest.fn(() => Promise.resolve({
		json: () => Promise.resolve({
			answer: "yes",
			forced: false,
			image: "https://yesno.wtf/assets/yes/2.gif",
		}),
	}));

	beforeEach(() => {
		wrapper = shallowMount(Indecision);
		clgSpy = jest.spyOn(console, "log");
		jest.clearAllMocks();
	});

	test("Debe de hacer match con el snapshot", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	test("Debe de no disparar nada al escribir en el input (console.log)", async () => {
		const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
		const input = wrapper.find("input");
		await input.setValue("Hola mundo");
		expect(clgSpy).toHaveBeenCalledTimes(1);
		expect(getAnswerSpy).not.toHaveBeenCalled();
	});

	test("Debe de disparar la acciÃ³n al escribir en el input con '?'", async () => {
		const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
		const input = wrapper.find("input");
		await input.setValue("Hola mundo?");
		expect(getAnswerSpy).toHaveBeenCalledTimes(1);
	});

	test("Debe de retornar una respuesta (getAnswer)", async () => {
		await wrapper.vm.getAnswer();
		const image = wrapper.find("img");
		expect(image.exists()).toBeTruthy();
		expect(wrapper.vm.image).toBe("https://yesno.wtf/assets/yes/2.gif");
		expect(wrapper.vm.answer).toBe("Si!");
	});

	test("Debe de retornar un fallo (getAnswer)", async () => {
		fetch.mockImplementationOnce(() => Promise.reject("Error"));
		await wrapper.vm.getAnswer();
		const image = wrapper.find("img");
		expect(image.exists()).toBeFalsy();
		expect(wrapper.vm.answer).toBe("Error");
	});
});
