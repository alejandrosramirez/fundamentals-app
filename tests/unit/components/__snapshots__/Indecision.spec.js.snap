// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Indecision /> Debe de hacer match con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Haz una pregunta...">
  <p>Recuerda terminar con un '?'</p>
  <!--v-if-->
</div>
`;
